foundry/auto_self_nominate.rs

```diff
@@ -0,0 +1,157 @@
+                    &defualt_metadata,
+                    target_deposite,
+                );
+                thread::sleep(Duration::from_millis(interval));
+            })
+            .unwrap();
+    }
+
+    fn send(
+        client: &Arc<dyn ConsensusClient>,
+        signer: &EngineSigner,
+        account_address: &PlatformAddress,
+        metadata: &str,
+        targetdep: u64,
+    ) {
+        let metabytes = metadata.rlp_bytes();
+        let mut dep = targetdep;
+        let address = account_address.address();
+        let block_id = BlockId::Latest;
+        let state = client.state_at(block_id).unwrap();
+
+        let banned = Banned::load_from_state(&state).unwrap();
+        if banned.is_banned(address) {
+            cwarn!(ENGINE, "Account is banned");
+            return
+        }
+
+        let jailed = Jail::load_from_state(&state).unwrap();
+        let prisoner = jailed.get_prisoner(&address).unwrap();
+        let current_term = client.current_term_id(block_id).unwrap();
+        if prisoner.custody_until <= (current_term + NEED_NOMINATION_UNDER_TERM_LEFT) {
+            cwarn!(ENGINE, "Account is still in custody");
+            return
+        }
+
+        let candidate = Candidates::load_from_state(&state).unwrap();
+        let candidate_need_nomination = candidate.get_candidate(&address).unwrap();
+        if !candidate_need_nomination.nomination_ends_at.lt(&current_term) {
+            cwarn!(ENGINE, "No need self nominate");
+            return
+        }
+        if !candidate_need_nomination.nomination_ends_at.min(current_term).lt(&current_term) {
+            cwarn!(ENGINE, "No need self nominate");
+            return
+        }
+
+        if candidate_need_nomination.deposit.lt(&targetdep) {
+            dep = targetdep.min(targetdep);
+        } else {
+            dep = 0 as u64;
+        }
+        AutoSelfNomination::self_nomination_transaction(&client, &signer, dep, metabytes);
+    }
+
+    fn self_nomination_transaction(
+        client: &Arc<dyn ConsensusClient>,
+        signer: &EngineSigner,
+        deposit: u64,
+        metadata: Bytes,
+    ) {
+        let network_id = client.network_id();
+        let seq = match signer.address() {
+            Some(address) => client.latest_seq(address),
+            None => {
+                cwarn!(ENGINE, "Signer was not assigned yet");
+                return
+            }
+        };
+        let selfnominate = SelfNominate {
+            deposit,
+            metadata,
+        };
+        let tx = Transaction {
+            seq,
+            fee: 0,
+            network_id,
+            action: Action::Custom {
+                handler_id: CUSTOM_ACTION_HANDLER_ID,
+                bytes: selfnominate.rlp_bytes(),
+            },
+        };
+        let signature = match signer.sign_ecdsa(*tx.hash()) {
+            Ok(signature) => signature,
+            Err(e) => {
+                cwarn!(ENGINE, "Could not sign the message:{}", e);
```

Comment:
Let's use `cerror` here. Singing must succeed. If it failed, maybe there is a bug in our code.